src/Workspaces/Remote/Core/SolutionAssetProvider.cs

```diff
@@ -29,20 +35,65 @@ public SolutionAssetProvider(HostWorkspaceServices services)
             _services = services;
         }
 
-        public ValueTask GetAssetsAsync(Stream outputStream, int scopeId, Checksum[] checksums, CancellationToken cancellationToken)
+        public async ValueTask GetAssetsAsync(PipeWriter pipeWriter, int scopeId, Checksum[] checksums, CancellationToken cancellationToken)
         {
-            // Complete RPC right away so the client can start reading from the stream.
-            // The fire-and forget task starts writing to the output stream and the client will read it until it reads all expected data.
-            _ = Task.Run(async () =>
+            var assetStorage = _services.GetRequiredService<ISolutionAssetStorageProvider>().AssetStorage;
+            var serializer = _services.GetRequiredService<ISerializerService>();
+
+            SolutionAsset? singleAsset = null;
+            IReadOnlyDictionary<Checksum, SolutionAsset>? assetMap = null;
+
+            if (checksums.Length == 1)
             {
-                using var writer = new ObjectWriter(outputStream, leaveOpen: false, cancellationToken);
+                singleAsset = (await assetStorage.GetAssetAsync(scopeId, checksums[0], cancellationToken).ConfigureAwait(false)) ?? SolutionAsset.Null;
+            }
+            else
+            {
+                assetMap = await assetStorage.GetAssetsAsync(scopeId, checksums, cancellationToken).ConfigureAwait(false);
+            }
+
+            // Work around the lack of async stream writing in ObjectWriter, which is required when writing to the RPC pipe.
+            // Run two tasks - the first synchronously writes to a local pipe and the second asynchronosly transfers the data to the RPC pipe.
+            //
+            // Configure the pipe to never block on write (waiting for the reader to read). This prevents deadlocks but might result in more
+            // (non-contiguous) memory allocated for the underlying buffers. The amount of memory is bounded by the total size of the serialized assets.
+            var localPipe = new Pipe(RemoteHostAssetSerialization.PipeOptionsWithUnlimitedWriterBuffer);
 
-                var assetStorage = _services.GetRequiredService<ISolutionAssetStorageProvider>().AssetStorage;
-                var serializer = _services.GetRequiredService<ISerializerService>();
-                await RemoteHostAssetSerialization.WriteDataAsync(writer, assetStorage, serializer, scopeId, checksums, cancellationToken).ConfigureAwait(false);
-            }, cancellationToken);
+            Task.Run(() =>
```

Comment:
💭 It makes me feel icky that these operations are part of an async method, but the life of this operation is not captured in the life of the returned task.

Reply:
That's explained here:

```C#
// Complete RPC one we send the initial piece of data and start waiting for the writer to send more,
// so the client can start reading from the stream.
```

Why would it be a problem? We just need to start the tasks that write to the pipe. The client will read and then finish the operation.